payloadinterceptor.go

```diff
@@ -59,29 +83,45 @@ func (s *streamingClientInterceptor) CloseRequest() error {
 func (s *streamingClientInterceptor) CloseResponse() error {
 	err := s.StreamingClientConn.CloseResponse()
 	s.mu.Lock()
+	hasClosed := !s.responseClosed && s.requestClosed
 	s.responseClosed = true
-	shouldCall := s.requestClosed && !s.onCloseCalled
-	if shouldCall {
-		s.onCloseCalled = true
-	}
 	s.mu.Unlock()
-	if shouldCall {
+	if hasClosed {
 		s.onClose()
 	}
 	return err
 }
 
+func (s *streamingClientInterceptor) onClose() {
+	header := s.ResponseHeader()
+	s.state.end(s.ctx, header)
+}
+
 type streamingHandlerInterceptor struct {
 	connect.StreamingHandlerConn
 
-	receive func(any, connect.StreamingHandlerConn) error
-	send    func(any, connect.StreamingHandlerConn) error
+	ctx   context.Context //nolint:containedctx
```

Comment:
I think you might not need this (and thus not need the nolint directive) if you actually tracked the span instead of the ctx. So if `state` had a reference to the span, you wouldn't need to hang on to the right ctx here, right?